### PR TITLE
[SDK-574] Change ResourceCryptoService

### PR DESCRIPTION
**Pull request by bitPogo**
## Description

This is a refactoring of the ResourceCryptoService. It gets rid of JavaRx and aligns the handling of the current resource types.
Also errors during the de-/encryption are now correctly propagated as D4LExceptions.

## Motivation and Context

There is a need to pull out JavaRx completely out of the SDK since it aims for KMP. Also the current implementation does something, which belongs in another place.
This is part of [SDK-574](https://gesundheitscloud.atlassian.net/browse/SDK-574)

## How is it being implemented?

As stated it removes JavaRx and dries out structural patterns in terms of the resource en-/decryption.
This is part of [SDK-574](https://gesundheitscloud.atlassian.net/browse/SDK-574) and depends on #153.

## How Has This Been Tested?

All new cases had been covered by unittests. Also the module tests needed alignment in terms of the resource encryption.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have updated the changelog accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: sdk-core/src/main/java/care/data4life/sdk/CryptoService.kt [hunk 1]
```diff
@@ -123,32 +123,26 @@ open class CryptoService : CryptoProtocol, CryptoContract.Service {
             }
     }
 
-    override fun encryptAndEncodeString(key: GCKey, data: String): Single<String> {
-        return encrypt(key, data.toByteArray())
+    override fun encryptAndEncodeByteArray(key: GCKey, data: ByteArray): Single<String> {
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/CryptoService.kt:126]: :+1:

File: sdk-core/src/main/java/care/data4life/sdk/CryptoService.kt [hunk 2]
```diff
@@ -179,13 +192,13 @@ open class CryptoService : CryptoProtocol, CryptoContract.Service {
 
     private fun convertExchangeKeyToGCKey(exchangeKey: ExchangeKey): Single<GCKey> {
         return Single.fromCallable { exchangeKey }
-            .map { exchKey ->
-                if (exchKey.getVersion() !== KeyVersion.VERSION_1) {
-                    throw (CryptoException.InvalidKeyVersion(exchKey.getVersion().value))
-                } else if (exchKey.type === KeyType.APP_PUBLIC_KEY || exchKey.type === KeyType.APP_PRIVATE_KEY) {
+            .map { exchangeKey ->
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/CryptoService.kt:195]: 👍

File: sdk-core/src/main/java/care/data4life/sdk/crypto/CryptoContract.kt [hunk 3]
```diff
@@ -61,15 +61,25 @@ class CryptoContract {
 
         fun encrypt(key: GCKey, data: ByteArray): Single<ByteArray>
         fun decrypt(key: GCKey, data: ByteArray): Single<ByteArray>
-        // TODO Change interface so it Arbitrary Data can use it directly
+
+        // ToDo move this into a convenience layer
```
Comment (wmontwe) [inline sdk-core/src/main/java/care/data4life/sdk/crypto/CryptoContract.kt:65]: Why TODO is now ToDo?

**Review by wmontwe — DISMISSED** [review]
👍

File: sdk-core/src/main/java/care/data4life/sdk/crypto/CryptoContract.kt (hunk 3 above)
Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/crypto/CryptoContract.kt]: You are right...I miss typed it...I will change it in a minute

Comment (bitPogo) [reply sdk-core/src/main/java/care/data4life/sdk/crypto/CryptoContract.kt]: done

**Review by wmontwe — APPROVED** [review]
👍

**Pull request merged**